sources/SilkTouch/Caching/FileSystemCacheProvider.cs

```diff
@@ -0,0 +1,243 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Silk.NET.SilkTouch.Caching;
+
+/// <summary>
+/// An implementation of <see cref="ICacheProvider"/> using the filesystem.
+/// </summary>
+public class FileSystemCacheProvider(ILogger<FileSystemCacheProvider> logger) : ICacheProvider
+{
+    private readonly ConcurrentDictionary<string, SemaphoreSlim?> _semaphores = new();
+    private readonly ConcurrentDictionary<string, (FileStream, ZipArchive)> _noHostZips = new();
+
+    internal static string CommonDir { get; set; } = Environment.CurrentDirectory;
+
+    /// <inheritdoc />
+    public async Task<(string Path, bool IsNew)?> GetDirectory(
+        string cacheKey,
+        CacheIntent intent,
+        CacheFlags flags
+    ) =>
+        await CoreGetDirectory(cacheKey, intent, intent == CacheIntent.ResolvedForeignInput, flags);
```

Comment:
why not return the task directly instead of using await here?

Reply:
Awaiting a async method is different than just returning the output. This is also standard practice for calling async methods as far as I'm aware.

Reply:
everything ive seen in bcl is if youre going to have the inner core method you just return directly to save the overhead of a second state machine, usually in the context of an outer method that does arg checking before forwarding to keep the arg checking out of the async though

Reply:
Awaiting is required to link the stacks correctly as I understand it, which impacts exception handling and will probably has some async2 implications in the future as well. If this were best practice, there’d probably be an optimisation in Roslyn and/or a ReSharper codefix to do essentially what you describe, which I don’t believe is the case(?)

Reply:
rider suggests it for me but that might not be standard idk